### PR TITLE
Add support for FusionOS 22 and 23

### PR DESCRIPTION
**Pull request by linhuali**
## Description
- This change added autoinstall support and necessary test updates for FusionOS 22 and 23. 
- Tested FusionOS 22 and 23 with sata/ide/pvscsi/nvme, vmxnet3/e1000e, efi/bios and all passed except known issues.

## Test Log
```
VM information:
+---------------------------------------------------------------------------------+
| Guest OS Distribution     | FusionOS 23.0.1 x86_64                              |
+---------------------------------------------------------------------------------+
| GUI Installed             | False                                               |
+---------------------------------------------------------------------------------+
| CloudInit Version         | 21.4                                                |
+---------------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                              |
+---------------------------------------------------------------------------------+
| VMTools Version           | 12.0.5 (build-19716617)                             |
+---------------------------------------------------------------------------------+
| Config Guest Id           | other5xLinux64Guest                                 |
+---------------------------------------------------------------------------------+
| GuestInfo Guest Id        | other5xLinux64Guest                                 |
+---------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Other 5.x or later Linux (64-bit)                   |
+---------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                          |
+---------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                                  |
|                           | bitness='64'                                        |
|                           | distroName='FusionOS'                               |
|                           | distroVersion='23.0.1'                              |
|                           | familyName='Linux'                                  |
|                           | kernelVersion='5.10.0-136.23.0.99.u37.fos23.x86_64' |
|                           | prettyName='FusionOS 23'                            |
+---------------------------------------------------------------------------------+

Test Results (Total: 26, Passed: 22, Skipped: 4, Elapsed Time: 01:44:49)
+--------------------------------------------------------------------------+
| ID | Name                                 |   Status         | Exec Time |
+--------------------------------------------------------------------------+
| 01 | deploy_vm_bios_ide_e1000e            |   Passed         | 00:14:02  |
| 02 | check_inbox_driver                   |   Passed         | 00:01:29  |
| 03 | ovt_verify_install                   |   Passed         | 00:05:04  |
| 04 | ovt_verify_status                    |   Passed         | 00:00:46  |
| 05 | vgauth_check_service                 |   Passed         | 00:00:53  |
| 06 | check_ip_address                     |   Passed         | 00:00:38  |
| 07 | check_os_fullname                    |   Passed         | 00:01:13  |
| 08 | stat_balloon                         |   Passed         | 00:00:45  |
| 09 | stat_hosttime                        |   Passed         | 00:00:42  |
| 10 | device_list                          |   Passed         | 00:02:36  |
| 11 | check_quiesce_snapshot_custom_script |   Passed         | 00:01:07  |
| 12 | memory_hot_add_basic                 |   Passed         | 00:05:58  |
| 13 | cpu_hot_add_basic                    |   Passed         | 00:05:24  |
| 14 | cpu_multicores_per_socket            |   Passed         | 00:10:04  |
| 15 | check_efi_firmware                   | * Not Applicable | 00:00:44  |
| 16 | secureboot_enable_disable            | * Not Applicable | 00:00:41  |
| 17 | e1000e_network_device_ops            |   Passed         | 00:03:56  |
| 18 | vmxnet3_network_device_ops           |   Passed         | 00:03:05  |
| 19 | pvrdma_network_device_ops            |   Passed         | 00:08:54  |
| 20 | paravirtual_vhba_device_ops          |   Passed         | 00:07:57  |
| 21 | lsilogic_vhba_device_ops             | * Not Supported  | 00:00:42  |
| 22 | lsilogicsas_vhba_device_ops          | * Not Supported  | 00:00:44  |
| 23 | sata_vhba_device_ops                 |   Passed         | 00:08:08  |
| 24 | nvme_vhba_device_ops                 |   Passed         | 00:07:56  |
| 25 | nvdimm_cold_add_remove               |   Passed         | 00:08:19  |
| 26 | ovt_verify_uninstall                 |   Passed         | 00:01:54  |
+--------------------------------------------------------------------------+
```

```
VM information:
+------------------------------------------------------------------------------------+
| Guest OS Distribution     | FusionOS 22.0.4 x86_64                                 |
+------------------------------------------------------------------------------------+
| GUI Installed             | True                                                   |
+------------------------------------------------------------------------------------+
| CloudInit Version         | 19.4                                                   |
+------------------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                                 |
+------------------------------------------------------------------------------------+
| VMTools Version           | 12.0.5 (build-19716617)                                |
+------------------------------------------------------------------------------------+
| Config Guest Id           | other4xLinux64Guest                                    |
+------------------------------------------------------------------------------------+
| GuestInfo Guest Id        | other4xLinux64Guest                                    |
+------------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Other 4.x Linux (64-bit)                               |
+------------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                             |
+------------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                                     |
|                           | bitness='64'                                           |
|                           | distroName='FusionOS'                                  |
|                           | distroVersion='22.0.4'                                 |
|                           | familyName='Linux'                                     |
|                           | kernelVersion='4.19.90-2212.2.0.0181.u89.fos22.x86_64' |
|                           | prettyName='FusionOS 22'                               |
+------------------------------------------------------------------------------------+

Test Results (Total: 26, Passed: 22, Skipped: 4, Elapsed Time: 02:20:12)
+--------------------------------------------------------------------------+
| ID | Name                                 |   Status         | Exec Time |
+--------------------------------------------------------------------------+
| 01 | deploy_vm_bios_ide_e1000e            |   Passed         | 00:29:52  |
| 02 | check_inbox_driver                   |   Passed         | 00:01:54  |
| 03 | ovt_verify_install                   |   Passed         | 00:04:29  |
| 04 | ovt_verify_status                    |   Passed         | 00:03:17  |
| 05 | vgauth_check_service                 |   Passed         | 00:01:15  |
| 06 | check_ip_address                     |   Passed         | 00:01:00  |
| 07 | check_os_fullname                    |   Passed         | 00:01:33  |
| 08 | stat_balloon                         |   Passed         | 00:00:57  |
| 09 | stat_hosttime                        |   Passed         | 00:00:55  |
| 10 | device_list                          |   Passed         | 00:03:16  |
| 11 | check_quiesce_snapshot_custom_script |   Passed         | 00:01:39  |
| 12 | memory_hot_add_basic                 |   Passed         | 00:07:33  |
| 13 | cpu_hot_add_basic                    |   Passed         | 00:06:51  |
| 14 | cpu_multicores_per_socket            |   Passed         | 00:11:09  |
| 15 | check_efi_firmware                   | * Not Applicable | 00:00:52  |
| 16 | secureboot_enable_disable            | * Not Applicable | 00:00:48  |
| 17 | e1000e_network_device_ops            |   Passed         | 00:04:52  |
| 18 | vmxnet3_network_device_ops           |   Passed         | 00:03:45  |
| 19 | pvrdma_network_device_ops            |   Passed         | 00:11:27  |
| 20 | paravirtual_vhba_device_ops          |   Passed         | 00:09:44  |
| 21 | lsilogic_vhba_device_ops             | * Not Supported  | 00:00:51  |
| 22 | lsilogicsas_vhba_device_ops          | * Not Supported  | 00:01:00  |
| 23 | sata_vhba_device_ops                 |   Passed         | 00:09:20  |
| 24 | nvme_vhba_device_ops                 |   Passed         | 00:08:56  |
| 25 | nvdimm_cold_add_remove               |   Passed         | 00:09:10  |
| 26 | ovt_verify_uninstall                 |   Passed         | 00:02:12  |
+--------------------------------------------------------------------------+
```

### CODE REVIEW

File: autoinstall/README.md [hunk 1]
```diff
@@ -23,6 +23,7 @@
 22. For openSUSE Leap 15.3 or later unattend auto-install, please use file openSUSE/15/autoinst.xml.
 23. For BCLinux 8.x unattend auto-install, please use file BCLinux/8/ks.cfg.
 24. For BCLinux-for-Euler 21.10 unattend auto-install, please use file BCLinux-for-Euler/21.10/ks.cfg.
+25. For FusionOS unattend auto-install, please use files under Fusion.
```
Comment (keirazhang) [inline autoinstall/README.md:26]: For FusionOS -> FusionOS 22 and 23

File: autoinstall/FusionOS/server_without_GUI/ks.cfg [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+{% if new_user is defined and new_user != 'root' %}
+# Add new user to sudoer
+echo '{{ new_user }} ALL=(ALL) NOPASSWD:ALL' >/etc/sudoers.d/{{ new_user }}
+# Enable autologin
+sed -i '/\[daemon\]/a AutomaticLogin={{ new_user }}' /etc/gdm/custom.conf
+sed -i '/\[daemon\]/a AutomaticLoginEnable=True' /etc/gdm/custom.conf
+{% endif %}
+echo '{{ autoinstall_complete_msg }}' >/dev/ttyS0
+%end
```
Comment (keirazhang) [inline autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: I am considering whether it is necessary to maintain two kickstart files for FusionOS. After all, it is rare to test RHEL-like guest OS without desktop environment. 2 kickstart files will increase our maintenance and regression testing efforts when we have changes about deploy_vm. Unless it is a must-have test scenario, I suggest we only keep one ks.cfg for FusionOS or any other new RHEL-like OS. We can put the ks.cfg under FusionOS/22/ks.cfg, which can be used by FusionOS 22 or later. When there is a new release which can't use FusionOS/22/ks.cfg, supposing FusionOS 30, we can add a new kickstart like FusionOS/30/ks.cfg at that time.

File: linux/deploy_vm/deploy_vm_from_iso.yml [hunk 3]
```diff
@@ -249,7 +250,7 @@
         vm_get_fullname_timeout: 600
       when:
         - unattend_install_conf is defined
-        - unattend_install_conf | lower is not match('.*(minimal|server_without_gui|bclinux-for-euler).*')
+        - unattend_install_conf | lower is not match('.*(minimal|server_without_gui|bclinux-for-euler|fusionos).*')
```
Comment (keirazhang) [inline linux/deploy_vm/deploy_vm_from_iso.yml:253]: Doesn't FusionOS have built-in open-vm-tools?

File: README.md [hunk 4]
```diff
@@ -65,33 +65,34 @@ This project supports below scenarios for end-to-end guest operating system vali
...
+| FreeBSD 13 and later                                   | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
+| Pardus 21.2 Server,XFCE Desktop and later              | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
+| openSUSE Leap 15.3 and later                           | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
+| BCLinux 8.x                                            | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
+| BCLinux-for-Euler 21.10                                | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
+| Red Hat Enterprise Linux CoreOS (RHCOS) 4.13 and later |                                  | :heavy_check_mark:       | :heavy_check_mark:                 |
+| FusionOS 22 and 23                                     | :heavy_check_mark:               |                          | :heavy_check_mark:                 |
```
Comment (keirazhang) [inline README.md:95]: FusionOS or Fusion OS?

File: linux/deploy_vm/deploy_vm_from_iso.yml (hunk 3 above)
Comment (linhuali) [reply linux/deploy_vm/deploy_vm_from_iso.yml:253]: No built-in ovt in FusionOS-22/23 and no related packages in iso. Linkup network interface after autoinstall but no ip until reboot.

File: README.md (hunk 4 above)
Comment (linhuali) [reply README.md:95]: FusionOS. Please check the official website: https://www.xfusion.com/cn/product/operating-system/fusionos

File: autoinstall/FusionOS/server_without_GUI/ks.cfg (hunk 2 above)
Comment (linhuali) [reply autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: FusionOS-23 doesn't support server_with_GUI.

Comment (keirazhang) [reply autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: Then it is better to change the path to FusionOS/22/ks.cfg and FusionOS/23/ks.cfg

File: README.md (hunk 4 above)
Comment (keirazhang) [reply README.md:95]: Got it. Thanks!

File: autoinstall/FusionOS/server_without_GUI/ks.cfg (hunk 2 above)
Comment (linhuali) [reply autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: FusionOS-22 works well with both server_with_GUI and server_without_GUI modes.

Comment (keirazhang) [reply autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: I don't think it conflicts. The unattended install config files are our recommendations. We can't cover all kind of installation varieties. If FusionOS 22 supports GUI, we recommend to use a ks.cfg with GUI installation. Or, you need to describe it more clearly in autoinstall/README.md: Fusion 22 can use server_with_GUI/ks.cfg and server_without_GUI/ks.cfg. But Fusion 23 can only use server_without_GUI/ks.cfg.

Comment (linhuali) [reply autoinstall/FusionOS/server_without_GUI/ks.cfg:91]: Fixed in autoinstall/README.md for FusionOS 22 and 23 separately.

**Review by keirazhang — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**